File: src/update_tickers.ts
// US Stock Ticker Update — TypeScript/Bun port of update_tickers.py.
//
// Fetches US stock tickers from NASDAQ's stock screener API (NYSE/NASDAQ/AMEX),
// supplements open/high/low/close from Yahoo Finance for the most recent
// completed session, matches an S&P 500 constituent list from Wikipedia, and
// writes market-cap-sorted CSVs (one daily snapshot per run, committed to git).
//
// The full historical SQL dump is NOT maintained here — it is rebuilt on demand
// from the git history of tickers/all.csv by src/gen_history_sql.ts.
//
// Run:  bun run src/update_tickers.ts

import {
  NASDAQ_URL,
  WIKIPEDIA_SP500_RAW_URL,
  HEADERS,
  YAHOO_WORKERS,
  HISTORY_COLUMNS,
  normalizeSymbol,
  parseMarketCap,
  parseNumber,
  parseIntField,
  pyFloatRepr,
  intRepr,
  csvField,
  writeCsv,
  mapLimit,
  fetchYahooChartRaw,
  utcDate,
  sleep,
  ensureDir,
} from "./lib.ts";

// Snapshot CSVs (all.csv, sp500.csv, top_*, by_industry/*) use the history
// columns WITHOUT the `date` column (date only lives in the on-demand
// history.sql built from git by src/gen_history_sql.ts).
const OUTPUT_COLUMNS = HISTORY_COLUMNS.filter((c) => c !== "date");

const SP500_SYMBOL_RE = /^\|\{\{(?:NyseSymbol|NasdaqSymbol|BZX link)\|([^}|]+)/gm;

interface Ticker {
  symbol: string;
  name: string;
  price: number | null;
  marketCap: number | null;
  volume: number | null;
  industry: string;
}

interface Ohlc {
  date: string;
  open: number | null;
  high: number | null;
  low: number | null;
  close: number;
}

// --- S&P 500 constituents from Wikipedia (raw wikitext) --------------------

async function fetchSp500Symbols(): Promise<string[]> {
  console.log("Fetching S&P 500 constituents from Wikipedia...");
  await sleep(500 + Math.random() * 1000);
  try {
    const response = await fetch(WIKIPEDIA_SP500_RAW_URL, {
      headers: HEADERS,
      signal: AbortSignal.timeout(30000),
    });
    if (!response.ok) throw new Error(`Wikipedia ${response.status}`);
    const content = await response.text();

    const startMarker = "== S&P 500 component stocks ==";
    const endMarker = "== Selected changes to the list of S&P 500 components ==";
    if (!content.includes(startMarker)) {
      throw new Error("Wikipedia page format changed: missing constituents section");
    }
    let section = content.split(startMarker)[1] ?? "";
    if (section.includes(endMarker)) section = section.split(endMarker)[0];

    const symbols: string[] = [];
    const seen = new Set<string>();
    for (const match of section.matchAll(SP500_SYMBOL_RE)) {
      const raw = match[1];
      const normalized = normalizeSymbol(raw);
      if (!normalized || seen.has(normalized)) continue;
      seen.add(normalized);
      symbols.push(raw.trim().toUpperCase());
    }
    if (symbols.length < 450) {
      throw new Error(`Wikipedia parse returned too few symbols: ${symbols.length}`);
    }
    console.log(`  Found ${symbols.length} S&P 500 constituent tickers`);
    return symbols;
  } catch (e) {
    console.log(`  Error: ${e}`);
    return [];
  }
}

// --- NASDAQ screener --------------------------------------------------------

async function fetchTickers(): Promise<{ us: Ticker[]; all: Ticker[] }> {
  console.log("Fetching tickers from NASDAQ...");
  await sleep(500 + Math.random() * 1000);
  try {
    const url = `${NASDAQ_URL}?tableonly=true&download=true`;
    const response = await fetch(url, {
      headers: HEADERS,
      signal: AbortSignal.timeout(30000),
    });
    if (!response.ok) throw new Error(`NASDAQ ${response.status}`);
    const payload: any = await response.json();
    const rows: any[] = payload?.data?.rows ?? [];
    if (!rows.length) {
      console.log("  No data returned from API");
      return { us: [], all: [] };
    }
    console.log(`  Fetched ${rows.length} tickers from API`);

    const all: Ticker[] = [];
    const us: Ticker[] = [];
    const seen = new Set<string>();
    let nonUs = 0;
    let duplicates = 0;

    for (const row of rows) {
      const symbol = String(row.symbol ?? "").trim();
      if (!symbol || seen.has(symbol)) {
        duplicates++;
        continue;
      }
      seen.add(symbol);
      const ticker: Ticker = {
        symbol,
        name: String(row.name ?? ""),
        price: parseNumber(row.lastsale ?? ""),
        marketCap: parseMarketCap(row.marketCap ?? ""),
        volume: parseIntField(row.volume ?? ""),
        industry: String(row.sector ?? ""),
      };
      all.push(ticker);
      if (row.country === "United States") us.push(ticker);
      else nonUs++;
    }

    console.log(`  Excluded: ${nonUs} non-US, ${duplicates} duplicates`);
    console.log(`  Found ${us.length} unique US tickers`);
    console.log(`  Found ${all.length} total unique listed tickers`);
    return { us, all };
  } catch (e) {
    console.log(`  Error: ${e}`);
    return { us: [], all: [] };
  }
}

// --- Yahoo OHLC (most recent completed session) -----------------------------

async function fetchYahooOhlc(symbol: string): Promise<Ohlc | null> {
  const result = await fetchYahooChartRaw(symbol, { range: "5d", interval: "1d" }, 10000);
  if (!result) return null;
  const timestamps: number[] = result.timestamp ?? [];
  const quote = result.indicators?.quote?.[0] ?? {};
  const opens: (number | null)[] = quote.open ?? [];
  const highs: (number | null)[] = quote.high ?? [];
  const lows: (number | null)[] = quote.low ?? [];
  const closes: (number | null)[] = quote.close ?? [];

  for (let i = timestamps.length - 1; i >= 0; i--) {
    const close = closes[i];
    if (close == null) continue;
    return {
      date: utcDate(timestamps[i]),
      open: opens[i] ?? null,
      high: highs[i] ?? null,
      low: lows[i] ?? null,
      close,
    };
  }
  return null;
}

async function fetchOhlcMap(symbols: Iterable<string>): Promise<Map<string, Ohlc>> {
  const list = Array.from(symbols);
  const total = list.length;
  console.log(`\nFetching OHLC for ${total} symbols from Yahoo Finance...`);
  console.log(`  (this is the slowest step; ~${total} requests)`);

  const ohlcMap = new Map<string, Ohlc>();
  await mapLimit(list, YAHOO_WORKERS, async (sym) => {
    const result = await fetchYahooOhlc(sym);
    if (result) ohlcMap.set(sym, result);
    return result;
  }, (done) => {
    if (done % 500 === 0 || done === total) {
      console.log(`  Progress: ${done}/${total} (${ohlcMap.size} with OHLC)`);
    }
  });

  console.log(`  Got OHLC for ${ohlcMap.size}/${total} symbols`);
  return ohlcMap;
}

// --- Filtering + file output -----------------------------------------------

function filterSp500Tickers(tickers: Ticker[], sp500Symbols: string[]): Ticker[] {
  if (!tickers.length || !sp500Symbols.length) return [];
  const sp500Map = new Map(sp500Symbols.map((s) => [normalizeSymbol(s), s]));
  const matched: Ticker[] = [];
  const unmatched = new Set(sp500Map.keys());

  for (const t of tickers) {
    const n = normalizeSymbol(t.symbol);
    if (sp500Map.has(n)) {
      matched.push(t);
      unmatched.delete(n);
    }
  }
  console.log(`  Matched ${matched.length} NASDAQ rows to S&P 500 constituents`);
  if (unmatched.size) {
    const sample = Array.from(unmatched).sort().slice(0, 10).join(", ");
    const suffix = unmatched.size > 10 ? "..." : "";
    console.log(`  Warning: ${unmatched.size} S&P symbols not found in NASDAQ data: ${sample}${suffix}`);
  }
  return matched;
}

/** Build one output row in the exact column order, attaching OHLC. */
function tickerOutputRow(t: Ticker, ohlcMap: Map<string, Ohlc>, industry: string): string[] {
  const o = ohlcMap.get(t.symbol) ?? null;
  return [
    t.symbol,
    t.name,
    pyFloatRepr(t.price),
    pyFloatRepr(o?.open ?? null),
    pyFloatRepr(o?.high ?? null),
    pyFloatRepr(o?.low ?? null),
    pyFloatRepr(o?.close ?? null),
    pyFloatRepr(t.marketCap),
    intRepr(t.volume),
    industry,
  ];
}

function sortByMarketCapDesc(tickers: Ticker[]): Ticker[] {
  return [...tickers].sort((a, b) => {
    const ma = a.marketCap === null ? -Infinity : a.marketCap;
    const mb = b.marketCap === null ? -Infinity : b.marketCap;
    return mb - ma;
  });
}

function safeIndustryFilename(industry: string): string {
  let safe = industry.toLowerCase().replace(/[^\w\s-]/g, "");
  safe = safe.trim().replace(/\s+/g, "_");
  return safe;
}

async function saveFiles(
  tickers: Ticker[],
  sp500Tickers: Ticker[] | null,
  ohlcMap: Map<string, Ohlc>,
): Promise<boolean> {
  if (!tickers.length) {
    console.log("No tickers to save!");
    return false;
  }

  // all.csv: normalize blank industry → Uncategorized, sort by market cap.
  const normalized = tickers.map((t) => ({
    t,
    industry: t.industry.trim() === "" ? "Uncategorized" : t.industry.trim(),
  }));
  const orderedUs = sortByMarketCapDesc(normalized.map((n) => n.t));
  const industryBySymbol = new Map(normalized.map((n) => [n.t.symbol, n.industry]));
  const allRows = orderedUs.map((t) => tickerOutputRow(t, ohlcMap, industryBySymbol.get(t.symbol) ?? "Uncategorized"));

  ensureDir("tickers");
  ensureDir("by_industry");

  console.log("\nSaving ticker lists...");
  await writeCsv("tickers/all.csv", [OUTPUT_COLUMNS, ...allRows]);
  console.log(`  - tickers/all.csv (${allRows.length} rows)`);

  if (sp500Tickers && sp500Tickers.length) {
    const spOrdered = sortByMarketCapDesc(sp500Tickers);
    const spRows = spOrdered.map((t) => tickerOutputRow(t, ohlcMap, t.industry));
    await writeCsv("tickers/sp500.csv", [OUTPUT_COLUMNS, ...spRows]);
    console.log(`  - tickers/sp500.csv (${spRows.length} rows)`);
  }

  for (const n of [50, 100, 200]) {
    await writeCsv(`tickers/top_${n}.csv`, [OUTPUT_COLUMNS, ...allRows.slice(0, n)]);
    console.log(`  - tickers/top_${n}.csv`);
  }

  console.log("\nSaving by industry...");
  // Group by industry in market-cap order, write one file per (alphabetical).
  const byIndustry = new Map<string, string[][]>();
  for (let i = 0; i < orderedUs.length; i++) {
    const ind = industryBySymbol.get(orderedUs[i].symbol) ?? "Uncategorized";
    if (!byIndustry.has(ind)) byIndustry.set(ind, []);
    byIndustry.get(ind)!.push(allRows[i]);
  }
  for (const industry of Array.from(byIndustry.keys()).sort()) {
    const rows = byIndustry.get(industry)!;
    const safe = safeIndustryFilename(industry);
    await writeCsv(`by_industry/${safe}.csv`, [OUTPUT_COLUMNS, ...rows]);
    console.log(`  - by_industry/${safe}.csv (${rows.length} tickers)`);
  }
  return true;
}

// --- main -------------------------------------------------------------------

async function main() {
  console.log("==================================================");
  console.log("US Stock Ticker Update");
  console.log("==================================================");

  const { us: tickers, all: allTickers } = await fetchTickers();
  const sp500Symbols = await fetchSp500Symbols();

  if (!tickers.length) {
    console.log("\nNo data found!");
    process.exit(1);
  }
  if (!sp500Symbols.length) {
    console.log("\nNo S&P 500 data found!");
    process.exit(1);
  }

  const sp500Tickers = filterSp500Tickers(allTickers, sp500Symbols);

  // OHLC for every symbol we will write out (US list + S&P 500 match).
  const ohlcSymbols = new Set<string>();
  for (const t of tickers) ohlcSymbols.add(t.symbol);
  for (const t of sp500Tickers) ohlcSymbols.add(t.symbol);
  const ohlcMap = await fetchOhlcMap(ohlcSymbols);

  if (await saveFiles(tickers, sp500Tickers, ohlcMap)) {
    console.log("\n==================================================");
    console.log("Update completed!");
    console.log("==================================================");
  } else {
    console.log("\nFailed to save!");
    process.exit(1);
  }
}

if (import.meta.main) {
  await main();
}

File: update_tickers.py
"""
US Stock Ticker Fetcher
=======================
Fetches US stock tickers from NASDAQ's stock screener API (covers NYSE, NASDAQ, AMEX),
then saves to CSV files sorted by market capitalization.
"""

import os
import re
import math
import random
import time
import requests
import pandas as pd
from datetime import datetime, timezone
from concurrent.futures import ThreadPoolExecutor, as_completed

# Configuration
NASDAQ_URL = "https://api.nasdaq.com/api/screener/stocks"
YAHOO_CHART_URL = "https://query1.finance.yahoo.com/v8/finance/chart/{}"
WIKIPEDIA_SP500_RAW_URL = "https://en.wikipedia.org/w/index.php?title=List_of_S%26P_500_companies&action=raw"
HEADERS = {
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
    "Accept": "application/json",
}

# How many symbols to fetch from Yahoo Finance in parallel. Yahoo's unofficial
# chart endpoint rate-limits aggressively, so keep this modest.
YAHOO_WORKERS = 15
SP500_SYMBOL_PATTERN = re.compile(
    r"^\|\{\{(?:NyseSymbol|NasdaqSymbol|BZX link)\|([^}|]+)",
    re.MULTILINE,
)


def normalize_symbol(symbol):
    """Normalize symbol formatting across data sources."""
    return symbol.strip().upper().replace(".", "/")


def fetch_sp500_symbols():
    """Fetch current S&P 500 constituent tickers from Wikipedia."""
    print("Fetching S&P 500 constituents from Wikipedia...")

    time.sleep(random.uniform(0.5, 1.5))

    try:
        response = requests.get(WIKIPEDIA_SP500_RAW_URL, headers=HEADERS, timeout=30)
        response.raise_for_status()

        content = response.text
        start_marker = "== S&P 500 component stocks =="
        end_marker = "== Selected changes to the list of S&P 500 components =="

        if start_marker not in content:
            raise ValueError("Wikipedia page format changed: missing constituents section")

        section = content.split(start_marker, 1)[1]
        if end_marker in section:
            section = section.split(end_marker, 1)[0]

        symbols = []
        seen = set()
        for symbol in SP500_SYMBOL_PATTERN.findall(section):
            normalized = normalize_symbol(symbol)
            if not normalized or normalized in seen:
                continue
            seen.add(normalized)
            symbols.append(symbol.strip().upper())

        if len(symbols) < 450:
            raise ValueError(f"Wikipedia parse returned too few symbols: {len(symbols)}")

        print(f"  Found {len(symbols)} S&P 500 constituent tickers")
        return symbols

    except Exception as e:
        print(f"  Error: {e}")
        return []


def fetch_tickers():
    """
    Fetches NASDAQ screener rows and splits them into all listings and US listings.
    """
    print("Fetching tickers from NASDAQ...")
    
    # Random initial delay (0.5-1.5s) to avoid predictable patterns
    time.sleep(random.uniform(0.5, 1.5))
    
    try:
        params = {
            "tableonly": "true",
            "download": "true"
        }
        
        response = requests.get(NASDAQ_URL, params=params, headers=HEADERS, timeout=30)
        response.raise_for_status()
        
        rows = response.json().get('data', {}).get('rows', [])
        
        if not rows:
            print("  No data returned from API")
            return []
        
        print(f"  Fetched {len(rows)} tickers from API")
        
        # Parse tickers with deduplication
        all_tickers = []
        us_tickers = []
        seen_symbols = set()
        non_us_count = 0
        duplicate_count = 0
        
        for row in rows:
            symbol = row.get('symbol', '').strip()
            
            # Skip duplicates
            if not symbol or symbol in seen_symbols:
                duplicate_count += 1
                continue
            seen_symbols.add(symbol)

            ticker = {
                'symbol': symbol,
                'name': row.get('name', ''),
                'price': parse_number(row.get('lastsale', '')),
                'marketCap': parse_market_cap(row.get('marketCap', '')),
                'volume': parse_int(row.get('volume', '')),
                'industry': row.get('sector', ''),
            }
            all_tickers.append(ticker)

            if row.get('country') == 'United States':
                us_tickers.append(ticker)
            else:
                non_us_count += 1
        
        print(f"  Excluded: {non_us_count} non-US, {duplicate_count} duplicates")
        print(f"  Found {len(us_tickers)} unique US tickers")
        print(f"  Found {len(all_tickers)} total unique listed tickers")
        return us_tickers, all_tickers
        
    except Exception as e:
        print(f"  Error: {e}")
        return [], []


def parse_market_cap(s):
    """Parse market cap string like '$1.2T' or '$500M' to numeric."""
    if not s or s == 'N/A':
        return None
    try:
        s = s.replace('$', '').replace(',', '').strip()
        multipliers = {'T': 1e12, 'B': 1e9, 'M': 1e6, 'K': 1e3}
        for suffix, mult in multipliers.items():
            if s.endswith(suffix):
                return float(s[:-1]) * mult
        return float(s)
    except:
        return None


def parse_number(s):
    """Parse price/number string to float."""
    if not s or s == 'N/A':
        return None
    try:
        return float(s.replace('$', '').replace(',', '').strip())
    except:
        return None


def parse_int(s):
    """Parse volume string to integer."""
    if not s or s == 'N/A':
        return None
    try:
        return int(str(s).replace(',', '').strip())
    except:
        return None


def yahoo_symbol(symbol):
    """Convert a NASDAQ symbol to the Yahoo Finance format.

    NASDAQ uses dots for share classes (e.g. BRK.B, BF.B); Yahoo uses dashes
    (BRK-B, BF-B). Some symbols use slashes; Yahoo also wants dashes there.
    """
    return symbol.strip().upper().replace(".", "-").replace("/", "-")


def fetch_yahoo_ohlc(symbol):
    """Fetch the most recent completed trading day's OHLC from Yahoo Finance.

    Uses a 5-day window so we still get the latest close when the script runs
    before US market open (10:00 UTC) and today's session has no data yet.
    Returns a dict {open, high, low, close} or None on failure/404.
    """
    url = YAHOO_CHART_URL.format(yahoo_symbol(symbol))
    params = {"range": "5d", "interval": "1d"}

    for attempt in range(3):
        try:
            time.sleep(random.uniform(0, 0.05))
            response = requests.get(url, params=params, headers=HEADERS, timeout=10)

            # Rate limited / temporarily blocked: back off and retry.
            if response.status_code in (401, 429) and attempt < 2:
                time.sleep(2 ** attempt)
                continue
            # Unknown symbol or delisted: no point retrying.
            if response.status_code == 404:
                return None
            response.raise_for_status()

            result = response.json().get("chart", {}).get("result", [None])[0]
            if not result:
                return None

            timestamps = result.get("timestamp") or []
            quote = (result.get("indicators") or {}).get("quote", [{}])[0]
            opens = quote.get("open", [])
            highs = quote.get("high", [])
            lows = quote.get("low", [])
            closes = quote.get("close", [])

            # Walk backwards to the most recent day that has a valid close.
            for i in range(len(timestamps) - 1, -1, -1):
                close = closes[i] if i < len(closes) else None
                if close is None:
                    continue
                trade_date = datetime.fromtimestamp(timestamps[i], tz=timezone.utc).date().isoformat()
                return {
                    "date": trade_date,
                    "open": opens[i] if i < len(opens) else None,
                    "high": highs[i] if i < len(highs) else None,
                    "low": lows[i] if i < len(lows) else None,
                    "close": close,
                }
            return None
        except Exception:
            if attempt < 2:
                time.sleep(2 ** attempt)
                continue
            return None


def fetch_ohlc_map(symbols):
    """Fetch latest-day OHLC for a set of symbols concurrently from Yahoo Finance.

    Best-effort: symbols that fail or are not found are simply left without
    OHLC (the columns will be blank in the CSV) rather than failing the run.
    """
    symbols = list(symbols)
    total = len(symbols)
    print(f"\nFetching OHLC for {total} symbols from Yahoo Finance...")
    print("  (this is the slowest step; ~{:d} requests)".format(total))

    ohlc_map = {}
    completed = 0

    with ThreadPoolExecutor(max_workers=YAHOO_WORKERS) as executor:
        futures = {executor.submit(fetch_yahoo_ohlc, s): s for s in symbols}
        for future in as_completed(futures):
            sym = futures[future]
            completed += 1
            try:
                result = future.result()
            except Exception:
                result = None
            if result:
                ohlc_map[sym] = result
            if completed % 500 == 0 or completed == total:
                print(f"  Progress: {completed}/{total} ({len(ohlc_map)} with OHLC)")

    print(f"  Got OHLC for {len(ohlc_map)}/{total} symbols")
    return ohlc_map


# Columns stored in the historical archive (CSV + SQL), in order.
HISTORY_COLUMNS = [
    "date", "symbol", "name", "price",
    "open", "high", "low", "close",
    "marketCap", "volume", "industry",
]
HISTORY_STRING_COLUMNS = {"date", "symbol", "name", "industry"}

HISTORY_CSV_PATH = "tickers/history.csv"
HISTORY_SQL_PATH = "tickers/history.sql"


def snapshot_date_from_ohlc(ohlc_map, fallback=None):
    """Pick the trading day this run represents.

    Each run captures the most recent *completed* US trading session (the
    workflow runs at 10:00 UTC, before US market open), so the snapshot date is
    the latest trade date returned by Yahoo across all symbols. Falls back to
    `fallback` (e.g. today's UTC date) when Yahoo returned nothing.
    """
    dates = [v.get("date") for v in ohlc_map.values() if v and v.get("date")]
    if dates:
        return max(dates)
    return fallback


def append_history_csv(tickers, ohlc_map, snapshot_date, path=HISTORY_CSV_PATH):
    """Append one dated cross-section of all US tickers to the history CSV.

    Idempotent: if the snapshot date is already present, the run is skipped,
    so re-running the workflow the same day (or running before the next session
    closes) does not create duplicate rows.
    """
    ohlc_map = ohlc_map or {}
    os.makedirs(os.path.dirname(path), exist_ok=True)

    # Idempotency: only the date column is needed to decide whether to skip.
    if os.path.exists(path):
        try:
            existing_dates = set(
                pd.read_csv(path, usecols=["date"])["date"].astype(str).unique()
            )
        except Exception:
            existing_dates = set()
        if snapshot_date in existing_dates:
            print(f"  history.csv already contains {snapshot_date}; skipping append")
            return False

    rows = []
    for t in tickers:
        o = ohlc_map.get(t.get("symbol")) or {}
        rows.append({
            "date": snapshot_date,
            "symbol": t.get("symbol"),
            "name": t.get("name"),
            "price": t.get("price"),
            "open": o.get("open"),
            "high": o.get("high"),
            "low": o.get("low"),
            "close": o.get("close"),
            "marketCap": t.get("marketCap"),
            "volume": t.get("volume"),
            "industry": t.get("industry"),
        })

    new_df = pd.DataFrame(rows, columns=HISTORY_COLUMNS)
    # Stable ordering within a day (largest market cap first) keeps the file tidy.
    new_df = new_df.sort_values("marketCap", ascending=False, na_position="last").reset_index(drop=True)

    write_header = not os.path.exists(path)
    new_df.to_csv(path, mode="a", header=write_header, index=False)
    print(f"  Appended {len(new_df)} rows for {snapshot_date} to {path}")
    return True


def _sql_literal(value, column):
    """Render a Python value as a SQL literal (portable across SQLite/PG/MySQL)."""
    if value is None:
        return "NULL"
    if isinstance(value, float) and math.isnan(value):
        return "NULL"
    if column in HISTORY_STRING_COLUMNS:
        text = "" if value is None else str(value)
        return "'" + text.replace("'", "''") + "'"
    return str(value)


def generate_history_sql(csv_path=HISTORY_CSV_PATH, sql_path=HISTORY_SQL_PATH):
    """Regenerate the SQL dump from the history CSV.

    One multi-row INSERT is emitted per trading day, which keeps import fast
    while staying well within statement-size limits. Output is deterministic so
    the committed file only diffs by the newly appended day.
    """
    if not os.path.exists(csv_path):
        print("  No history.csv found; skipping SQL generation")
        return False

    df = pd.read_csv(csv_path, dtype={"date": str})
    # Keep NULLs as None; sort for stable, grouped output.
    df = df.sort_values(["date", "marketCap"], ascending=[True, False], na_position="last")

    lines = [
        "-- US stock ticker history (auto-generated daily by update_tickers.py).",
        "-- Columns: date = trading day of the OHLC bar; price/marketCap/volume",
        "-- are the NASDAQ screener snapshot taken during the same run.",
        "-- Portable SQL: works on SQLite, PostgreSQL, and MySQL.",
        "",
        "CREATE TABLE IF NOT EXISTS us_tickers (",
        "    date      DATE        NOT NULL,",
        "    symbol    VARCHAR(32) NOT NULL,",
        "    name      VARCHAR(255),",
        "    price     DECIMAL(18,4),",
        "    open      DECIMAL(18,4),",
        "    high      DECIMAL(18,4),",
        "    low       DECIMAL(18,4),",
        "    close     DECIMAL(18,4),",
        "    marketCap DECIMAL(24,2),",
        "    volume    BIGINT,",
        "    industry  VARCHAR(128),",
        "    PRIMARY KEY (date, symbol)",
        ");",
        "",
    ]

    cols = HISTORY_COLUMNS
    total_rows = 0
    for date_val, group in df.groupby("date", sort=True):
        lines.append(
            "INSERT INTO us_tickers (" + ", ".join(cols) + ") VALUES"
        )
        value_rows = []
        for record in group[cols].itertuples(index=False, name=None):
            literals = [_sql_literal(v, c) for v, c in zip(record, cols)]
            value_rows.append("    (" + ", ".join(literals) + ")")
            total_rows += 1
        lines.append(",\n".join(value_rows) + ";")
        lines.append("")

    with open(sql_path, "w", encoding="utf-8") as f:
        f.write("\n".join(lines))

    print(f"  Wrote {sql_path} ({total_rows} rows across {df['date'].nunique()} dates)")
    return True


def filter_sp500_tickers(tickers, sp500_symbols):
    """Filter ticker rows down to S&P 500 constituents."""
    if not tickers or not sp500_symbols:
        return []

    sp500_map = {normalize_symbol(symbol): symbol for symbol in sp500_symbols}
    matched_tickers = []
    unmatched_symbols = set(sp500_map)

    for ticker in tickers:
        normalized = normalize_symbol(ticker.get('symbol', ''))
        if normalized in sp500_map:
            matched_tickers.append(ticker)
            unmatched_symbols.discard(normalized)

    print(f"  Matched {len(matched_tickers)} NASDAQ rows to S&P 500 constituents")
    if unmatched_symbols:
        sample = ", ".join(sorted(unmatched_symbols)[:10])
        suffix = "..." if len(unmatched_symbols) > 10 else ""
        print(f"  Warning: {len(unmatched_symbols)} S&P symbols not found in NASDAQ data: {sample}{suffix}")

    return matched_tickers


def attach_ohlc(df, ohlc_map):
    """Attach open/high/low/close columns to a ticker DataFrame and order them
    right after price. Symbols missing from ohlc_map get blank OHLC cells."""
    ohlc_map = ohlc_map or {}
    for col in ["open", "high", "low", "close"]:
        df[col] = df["symbol"].map(lambda s, c=col: (ohlc_map.get(s) or {}).get(c))

    desired = [
        "symbol", "name", "price",
        "open", "high", "low", "close",
        "marketCap", "volume", "industry",
    ]
    return df[[c for c in desired if c in df.columns]]


def save_files(tickers, sp500_tickers=None, ohlc_map=None):
    """Save tickers to organized folder structure."""
    if not tickers:
        print("No tickers to save!")
        return False

    # Sort by market cap (descending)
    df = pd.DataFrame(tickers)
    df['industry'] = df['industry'].fillna('').astype(str).str.strip()
    df.loc[df['industry'] == '', 'industry'] = 'Uncategorized'
    df = df.sort_values('marketCap', ascending=False).reset_index(drop=True)
    df = attach_ohlc(df, ohlc_map)

    # Create output directories
    os.makedirs("tickers", exist_ok=True)
    os.makedirs("by_industry", exist_ok=True)

    # === GENERAL TICKER LISTS ===
    print("\nSaving ticker lists...")

    df.to_csv("tickers/all.csv", index=False)
    print(f"  - tickers/all.csv ({len(df)} rows)")

    if sp500_tickers:
        sp500_df = pd.DataFrame(sp500_tickers)
        sp500_df = sp500_df.sort_values('marketCap', ascending=False).reset_index(drop=True)
        sp500_df = attach_ohlc(sp500_df, ohlc_map)
        sp500_df.to_csv("tickers/sp500.csv", index=False)
        print(f"  - tickers/sp500.csv ({len(sp500_df)} rows)")

    for n in [50, 100, 200]:
        df.head(n).to_csv(f"tickers/top_{n}.csv", index=False)
        print(f"  - tickers/top_{n}.csv")
    
    # === BY INDUSTRY ===
    print("\nSaving by industry...")
    
    industries = df['industry'].unique()
    
    for industry in sorted(industries):
        industry_df = df[df['industry'] == industry]
        if len(industry_df) == 0:
            continue
        
        # Safe filename
        safe_name = re.sub(r'[^\w\s-]', '', industry.lower())
        safe_name = re.sub(r'\s+', '_', safe_name.strip())
        
        industry_df.to_csv(f"by_industry/{safe_name}.csv", index=False)
        print(f"  - by_industry/{safe_name}.csv ({len(industry_df)} tickers)")
    
    return True


if __name__ == "__main__":
    print("=" * 50)
    print("US Stock Ticker Update")
    print("=" * 50)
    
    tickers, all_tickers = fetch_tickers()
    sp500_symbols = fetch_sp500_symbols()
    
    if not tickers:
        print("\nNo data found!")
        exit(1)

    if not sp500_symbols:
        print("\nNo S&P 500 data found!")
        exit(1)

    sp500_tickers = filter_sp500_tickers(all_tickers, sp500_symbols)

    # Fetch latest-day OHLC for every symbol we will write out (US list plus
    # the S&P 500 match). The NASDAQ screener does not provide open/high/low,
    # so we supplement it from Yahoo Finance.
    ohlc_symbols = {t.get("symbol") for t in tickers} | {t.get("symbol") for t in sp500_tickers}
    ohlc_map = fetch_ohlc_map(ohlc_symbols)

    # Determine the trading day this run represents (most recent completed US
    # session), then append a dated cross-section to the history archive and
    # regenerate the importable SQL dump from it.
    snapshot_date = snapshot_date_from_ohlc(
        ohlc_map, fallback=datetime.now(timezone.utc).date().isoformat()
    )
    print(f"\nSnapshot date for history archive: {snapshot_date}")

    print("\nUpdating history archive...")
    append_history_csv(tickers, ohlc_map, snapshot_date)
    generate_history_sql()

    if save_files(tickers, sp500_tickers, ohlc_map):
        print("\n" + "=" * 50)
        print("Update completed!")
        print("=" * 50)
    else:
        print("\nFailed to save!")
        exit(1)
